### ISSUE TITLE
Preallocate files

### ISSUE THREAD

**Issue by reallyuniquename** [sabnzbd/sabnzbd#2169]
Right now missing articles are simply getting skipped. Remaining data is misaligned and resulting filesize is different. 

This behaviour complicates a lot of simple ways of restoring missing chunks even if you can get them from other sources. You can't easily fill just the missing data from P2P networks. You can't use RAR's internal recovery record. You can't unpack existing data from multipart archives even when total number of lost articles is relatively small.

I know, I know. NZB files do not provide length of actual data, only a number of raw bytes. But you can still guess and I believe some clients already do that quite fine.

I've just had to pull 50GB of data once again while I was missing less than 1MB. This is really frustrating. I also see an increasing number of uploaders that intentionally do not provide any PARs and 1 missing article trashes the whole batch.

Anyway this has been brought up a few times on forums with a lof valid points but I don't think it got any traction: `/viewtopic.php?t=9851` & `/viewtopic.php?t=15373`.

**Comment by Safihre** [comment]
It's easily possible, the information is inside the yEnc header. That's how nzbget does it. I already made a proof of concept a year or so ago. 
It's just quite a big investment and there was no real good reason to do it. Par2 repair works just as well in my tests with filled or unfilled data.
I understand your use case, but it's also a bit exotic.

A much better use case would be that this way we can actually implement a proper Retry where only the missing articles are tried again and filled in at the right spot in the file. That's the reason why nzbget implemented it.

**Comment by thezoggy** [comment]
there is tons of obfuscation methods, purposely not including pars / leaving stuff incomplete on purpose (force you to get the actual nzb from the soruce) and many other reasons. some are to avoid dmca some are to avoid other sites from leeching their stuff.

writing files is slow and puts wear on hdd.. why do it with 0 then to just do it again with said data... i dont see any benefit, just more wear on a hdd and wasting time?

**Comment by Safihre** [comment]
@thezoggy you're not actually writing data, you say `fp.seek(X)`. All the space between `0` and `X` will just not be filled with anything. So no extra writes.
Depends on the OS how it actually fills the space as far as I understand.

**Comment by reallyuniquename** [comment]
True, implementing this would make proper Retry possible too.  

>the information is inside the yEnc header

I'm not familiar with NNTP. If article is missing where would the header come from? 

@thezoggy 
It's not obfuscation, uploaders explicitly state they don't use PAR files. Those mostly come from one place but don't focus much on that. It's more about extracting existing data and its structure. I've had enough failed downloads with proper PAR sets that couldn't be salvaged despite having most of the articles.

As for the zero prefill it doesn't write anything if implemented properly, OS handles that. I mean full preallocation is not required to fill missing articles but I felt like it's a very related issue.

**Comment by Safihre** [comment]
If there's no article, indeed we can't parse the header.

Does it really need to preallocate the whole file? Why not add as we go? As long as the data is in the right location inside the file.
For example if the file is 100Kb, but the last article of 10Kb is missing, it's not fine to just leave it at 90Kb?

**Comment by reallyuniquename** [comment]
No, full instant preallocation is not required. I guess I shouldn't have mixed these two features in one issue. But they both require accounting for the size of missing articles.

I'm guessing most upload tools use fixed data length for every post in a single NZB. Can we use that to calculate correct size when article is missing? Would that work?

**Comment by animetosho** [comment]
> Does it really need to preallocate the whole file? Why not add as we go?

There are benefits with pre-allocation. Exact behaviour can vary across filesystems, but things include:

* space is preallocated, so you won't get 'out of space' errors during download. Also helps other programs be aware of the intended disk usage, so they won't get unexpected 'out of space' errors either
* reduces fragmentation
* I generally find pre-allocation to be faster than a filesystem constantly trying to increase the size of a file (and [calls like this](https://linux.die.net/man/3/posix_fallocate) tend to be fast)
* PAR2 repair is slightly more effective when data is in the right place. PAR2 can deal with misaligned data, but there's some minor efficiency loss, plus the mechanism it uses to do this is CRC rolling, which being a bit slow, many clients limit its usage
* the file can be used across other tools, e.g. imported into a torrent client to fetch missing parts

IMO, pre-allocating files is the most sensible approach for a downloader, if possible.

**Comment by reallyuniquename** [comment]
If preallocation gets a go please make sure to use instant file initialization using `SeManageVolumePrivilege` on Windows. Otherwise you'd be literally writing zeros.

IMO, this and filling missing articles should be separate options.

**Comment by Safihre** [comment]
It's interesting, but as I wrote before I don't see many convincing points (e.g. that would benefit 50%+ of users) to implement it right now.
Mayhe in the future.

**Comment by animetosho** [comment]
> If preallocation gets a go please make sure to use instant file initialization

Across other applications, I see it often being an option between no preallocation, fast preallocation (using OS supplied calls, or maybe seeking to the last byte and writing a 0) and "full" preallocation (explicitly zero-fill the file), with the default being on fast preallocation.  
Unfortunately, filesystems aren't consistent with how they operate. Particularly if the filesystem doesn't support sparse files, the OS may want to zero-fill the file for security reasons.

Of course, whether it's worth implementing is another judgement altogether. If I were implementing the system from scratch, I'd definitely take the approach, but changes on an existing system is a different cost.

**Comment by reallyuniquename** [comment]
Sparse files are usually what you want to stay away from. At least on conventional spinning rust. Insane fragmentation, extremely slow to read from. 

>I see it often being an option between no preallocation, fast preallocation (using OS supplied calls, or maybe seeking to the last byte and writing a 0) and "full" preallocation (explicitly zero-fill the file), with the default being on fast preallocation.

I've yet to see an app that transparently explains this to an end user. qBittorrent still takes the heat for stalling while it writes gigabytes of zeros on Windows for hours with preallocation enabled. The reason for that is libtorrent call to `SetFileValidData()` which allocates space without zeroing it requires additional permissions it never asks user for.

**Comment by puzzledsab** [comment]
This sparse file thing is much harder than I hoped. I modified newswrapper to extract ybegin and yend and tried to use it with fseek in assemble like some examples seemed to imply that I could. Unfortunately it creates broken files when I dump all available parts in every loop. Apparently you can't update data inside a file.

I don't think the files being generated this way are sparse at all or if it would work this way if they were. If they were, what happens to the data that crosses sectors? I think maybe we would have to write data in 4 KB blocks and join sections of parts to make sure they fit properly. I'm using Windows and NTFS.

I have verified that the ybegin value I extract for each part is correct (after adjusting for yenc kinks) by comparing it to `fout.tell()` when I'm using the current append mode and that the size calculated using ybegin and yend is the same as `length(data)`.

Related to #2459

**Comment by mnightingale** [comment]
Hmm python has strange open modes, I can't tell which would be "open in binary for writing without truncating"
Looking at https://docs.python.org/3.8/library/functions.html#open maybe "r+b".

Did you try that?

**Comment by mnightingale** [comment]
Also regarding the 4 KB writing thing, it might actually be worth investigating turning the buffering=0 option back to the default which will handle writing in optimal sizes, or at least understanding whether it’s ever advantageous to disable buffering.

**Comment by Safihre** [comment]
@mnightingale please take a look at the actual implementation in cpython of the buffering, I did and it's much more basic than you might think. There is barely anything smart about it, and the buffering limit is just a constant. Setting buffering=0 gives us a direct filepointer instead of all the useless overhead of buffering logic that we don't need because we always write blocks much bigger (750kb) than the buffer.

**Comment by animetosho** [comment]
@puzzledsab Do you have some code we can look at?

> If they were, what happens to the data that crosses sectors?

The filesystem APIs are supposed to hide the notion of sectors, so you shouldn't have to worry about that. Essentially, the data will get placed accordingly and split across sectors if necessary.

**Comment by puzzledsab** [comment]
@mnightingale : I did try it but apparently not hard enough. It seems to work now.

**Comment by mnightingale** [comment]
This feature request covers two things:

- Write data to correct offsets
- Preallocate files

The direct write feature achieves the first, but we are using sparse files rather than allocating physical space for files. This has the benefit that like append mode missing articles don’t take up physical disk space.
Actual preallocation would have the effect of reserving physical disk space, but it likely adds a lot of unnecessary complexity, security issues on some platforms, and is unlikely to provide any meaningful benefit.

**Linux** would be the simplest to implement since it can use fallocate or posix_fallocate.
**Windows** SetFileValidData which requires SeManageVolumePrivilege (requires run as admin?) and could introduce security issues.
**macOS** closest appears to be fcntl and F_PREALLOCATE

Space should be reserved instantly, but if they fail or are not supported an implementation may have to fallback to writing zeros which has caused problems with other software so I would not suggest it.

Except maybe Linux (fallocate or posix_fallocate) without a fallback and behind a special config option (disabled by default) I don’t think it’s worth pursing, especially with SSDs being so prevalent the fragmentation aspect is no longer a concern.

**Comment by Safihre** [comment]
I don't see the need for Preallocate, since it will also confuse users why files are already the size they expect.
Will close this now, since we have direct write!
